### PR TITLE
Refactoring/74 refactor GitHub api adapter

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**

### CODE REVIEW

File: src/main/java/com/exasol/releaserobot/main/ReleaseRobot.java [hunk 1]
```diff
@@ -28,10 +29,12 @@ public ReleaseRobot(final ReleaseUseCase releaseUseCase, final ValidateUseCase v
 
     /**
      * Main entry point for all Release Robot's calls.
+     *
+     * @throws GitHubException
      */
     // [impl->dsn~rr-starts-release-only-if-all-validation-succeed~1]
     // [impl->dsn~rr-runs-release-goal~1]
-    public void run(final UserInput userInput) {
+    public void run(final UserInput userInput) throws GitHubException {
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/main/ReleaseRobot.java:37]: this exception comes from the validate/release execution, and should be handled there, and be reflected in the reports. I would suggest leave it as it is and create a ticket for it.

File: src/main/java/com/exasol/releaserobot/repository/maven/JavaMavenRepository.java [hunk 2]
```diff
@@ -6,24 +6,28 @@
...
 
-import com.exasol.releaserobot.repository.*;
+import com.exasol.releaserobot.repository.GitRepositoryException;
+import com.exasol.releaserobot.usecases.Repository;
 
 /**
  * This class represents a Maven-based Java project's content.
  */
-public class JavaMavenGitBranch extends AbstractGitHubGitBranch {
+public class JavaMavenRepository extends Repository {
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/repository/maven/JavaMavenRepository.java:15]: what is this a `JavaMavenRepository` and not just a `MavenRepository` ? is there something Java-related I'm missing?

File: src/main/java/com/exasol/releaserobot/usecases/Repository.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import org.kohsuke.github.*;
+
+import com.exasol.releaserobot.repository.*;
+
+/**
+ * This class represents Git repository content based on the latest commit of the user-specified branch.
+ */
+public abstract class Repository {
+    private static final String CHANGELOG_FILE_PATH = "doc/changes/changelog.md";
+    private final GHRepository ghRepository;
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/usecases/Repository.java:15]: This should not be here. This is your internal model and should not know anything about Github.

The `RepositoryGateway` should give you a back valid `Repository` and make the right instantiation/transformation.

File: src/main/java/com/exasol/releaserobot/usecases/UserInput.java [hunk 4]
```diff
@@ -191,4 +191,8 @@ private void throwExceptionForMissingParameter(final String exceptionCode, final
             throw new IllegalArgumentException(message);
         }
     }
+
+    public String getRepositoryFullName() {
+        return this.getRepositoryOwner() + "/" + this.repositoryName;
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/usecases/UserInput.java:196]: Given that this is the only use of `RepositoryOwner`, I wonder if it should exists at all. Maybe we can remove it, and deal only with `repositoryName`, and users should specify the full repository name in the command line.

File: src/main/java/com/exasol/releaserobot/usecases/release/ReleaseInteractor.java [hunk 5]
```diff
@@ -33,24 +34,26 @@ public ReleaseInteractor(final ValidateUseCase validateUseCase, final Map<Platfo
...
     }
 
-    private Report makeRelease(final Set<PlatformName> platformNames) {
+    private Report makeRelease(final String repositoryFullName, final Set<PlatformName> platformNames) {
         final Report report = ReportImpl.releaseReport();
         for (final PlatformName platformName : platformNames) {
             try {
-                this.getReleaseMaker(platformName).makeRelease(this.repositoryGateway.getDefaultBranch());
+                this.getReleaseMaker(platformName)
+                        .makeRelease(this.repositoryGateway.getRepositoryWithDefaultBranch(repositoryFullName));
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/usecases/release/ReleaseInteractor.java:56]: getting the repository should be out of the loop. Right now we are getting it in every iteration

File: src/main/java/com/exasol/releaserobot/usecases/validate/PlatformValidator.java [hunk 6]
```diff
@@ -11,7 +11,6 @@
      * Validate a project.
      *
      * @param repository instance of {@link Repository}
-     *
      * @return validation report
      */
     public Report validate(final Repository repository);
```
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/releaserobot/usecases/validate/PlatformValidator.java:16]: we can remove this interface as it does exactly the same as `RepositoryValidator`

File: src/test/java/com/exasol/releaserobot/repository/GitHubGitRepositoryTest.java [hunk 7]
```diff
@@ -9,16 +9,8 @@
 import org.junit.jupiter.api.Test;
 import org.mockito.Mockito;
 
+import com.exasol.releaserobot.usecases.Repository;
+
 class GitHubGitRepositoryTest {
-    @Test
-    void testGitRepository() {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/releaserobot/repository/GitHubGitRepositoryTest.java:14]: why is this test removed?

File: src/test/java/com/exasol/releaserobot/repository/maven/JavaMavenGitBranchContentTest.java [hunk 8]
```diff
@@ -21,8 +22,8 @@
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 
-import com.exasol.releaserobot.repository.Branch;
 import com.exasol.releaserobot.repository.GitRepositoryException;
+import com.exasol.releaserobot.usecases.Repository;
 
 @ExtendWith(MockitoExtension.class)
 class JavaMavenGitBranchContentTest {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/releaserobot/repository/maven/JavaMavenGitBranchContentTest.java:29]: the test name does not match with the class now, right?

**Review by chiaradiamarcelo — CHANGES_REQUESTED** [review]
Looks good, I just left some comments

File: src/main/java/com/exasol/releaserobot/github/GithubRepositoryGateway.java [hunk 9]
```diff
@@ -14,30 +12,24 @@
...
 
     @Override
-    public Repository getRepository(final UserInput userInput) {
-        final Branch branch = this.getBranch(userInput);
-        final Optional<String> latestTag = this.githubGateway.getLatestTag();
-        return new GitHubGitRepository(latestTag, branch);
-    }
-
-    private Branch getBranch(final UserInput userInput) {
+    public Repository getRepository(final UserInput userInput) throws GitHubException {
```
[the remark this answers is not in the document]
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/github/GithubRepositoryGateway.java:23]: Renamed

File: src/main/java/com/exasol/releaserobot/main/ReleaseRobot.java (hunk 1 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/main/ReleaseRobot.java:37]: Created a ticket: https://github.com/exasol/release-robot/issues/76

File: src/main/java/com/exasol/releaserobot/repository/maven/JavaMavenRepository.java (hunk 2 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/repository/maven/JavaMavenRepository.java:15]: No,it's just a name. I renamed it

File: src/main/java/com/exasol/releaserobot/usecases/Repository.java (hunk 3 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/usecases/Repository.java:15]: It retrieves files as strings that I need for some checks. Should we store all the files in an object?

File: src/main/java/com/exasol/releaserobot/usecases/UserInput.java (hunk 4 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/usecases/UserInput.java:196]: the owner is hardcoded now (exasol), and the user doesn't  input it at all for convenience

File: src/main/java/com/exasol/releaserobot/usecases/validate/PlatformValidator.java (hunk 6 above)
Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/releaserobot/usecases/validate/PlatformValidator.java:16]: Removed it

File: src/test/java/com/exasol/releaserobot/repository/GitHubGitRepositoryTest.java (hunk 7 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/releaserobot/repository/GitHubGitRepositoryTest.java:14]: The logic was moved from the class

File: src/test/java/com/exasol/releaserobot/repository/maven/JavaMavenGitBranchContentTest.java (hunk 8 above)
Comment (AnastasiiaSergienko) [reply src/test/java/com/exasol/releaserobot/repository/maven/JavaMavenGitBranchContentTest.java:29]: fixed it

File: src/main/java/com/exasol/releaserobot/usecases/Repository.java (hunk 3 above)
Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/releaserobot/usecases/Repository.java:15]: We agreed to do it in another PR

File: src/main/java/com/exasol/releaserobot/usecases/UserInput.java (hunk 4 above)
Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/releaserobot/usecases/UserInput.java:196]: We agreed to do it in another PR

**Review by chiaradiamarcelo — APPROVED** [review]

**Pull request merged**